regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/VerboseRegexFinder.java

```diff
@@ -0,0 +1,167 @@
+        String raw = subElement.getText();
+        if (subElement.is(CHARACTER_RANGE)) {
+          hasDigit |= "0-9".equals(raw);
+          hasLowerCase |= "a-z".equals(raw);
+          hasUpperCase |= "A-Z".equals(raw);
+        } else if (subElement.is(PLAIN_CHARACTER)) {
+          hasUnderscore |= "_".equals(raw);
+        }
+      }
+
+      if (hasDigit && hasLowerCase && hasUpperCase && hasUnderscore) {
+        String expected = backslash(tree) + (tree.isNegated() ? "W" : "w");
+        reportVerboseCharacterClass(expected, tree);
+      }
+    }
+  }
+
+  private void checkBulkyNumericCharacterClass(CharacterClassTree tree) {
+    CharacterClassElementTree element = tree.getContents();
+    if (element.is(CHARACTER_RANGE) && "0-9".equals(element.getText())) {
+      String expected = backslash(tree) + (tree.isNegated() ? "D" : "d");
+      reportVerboseCharacterClass(expected, tree);
+    }
+  }
+
+  private void checkBulkyAnyCharacterClass(CharacterClassTree tree) {
+    CharacterClassElementTree element = tree.getContents();
+    if (tree.isNegated() || !element.is(UNION) || ((CharacterClassUnionTree) element).getCharacterClasses().size() != 2) {
+      return;
+    }
+
+    boolean hasLowerEscapeW = false;
+    boolean hasUpperEscapeW = false;
+    boolean hasLowerEscapeD = false;
+    boolean hasUpperEscapeD = false;
+    boolean hasLowerEscapeS = false;
+    boolean hasUpperEscapeS = false;
+
+    for (CharacterClassElementTree subElement : ((CharacterClassUnionTree) element).getCharacterClasses()) {
+      if (subElement.is(ESCAPED_CHARACTER_CLASS)) {
+        String raw = subElement.getText();
+        hasLowerEscapeW |= "\\w".equals(raw);
+        hasUpperEscapeW |= "\\W".equals(raw);
+        hasLowerEscapeD |= "\\d".equals(raw);
+        hasUpperEscapeD |= "\\D".equals(raw);
+        hasLowerEscapeS |= "\\s".equals(raw);
+        hasUpperEscapeS |= "\\S".equals(raw);
+      }
+    }
+
+    boolean isBulkyAnyCharacterClass = (hasLowerEscapeW && hasUpperEscapeW)
+      || (hasLowerEscapeD && hasUpperEscapeD)
+      || (hasLowerEscapeS && hasUpperEscapeS && tree.activeFlags().contains(Pattern.DOTALL));
+    if (isBulkyAnyCharacterClass) {
+      reportVerboseCharacterClass(".", tree);
+    }
+  }
+
+  private void reportVerboseCharacterClass(String expected, CharacterClassTree tree) {
+    String message = String.format(VERBOSE_CHARACTER_CLASS_MESSAGE, expected, tree.getText());
+    regexElementIssueReporter.report(tree, message, null, Collections.emptyList());
+  }
+
+  private void checkBulkyQuantifier(Quantifier quantifier) {
+    String raw = quantifier.getText();
+    BulkyQuantifier bulkyQuantifier = null;
+    if (Pattern.matches("\\{0,1}\\??$", raw)) {
+      bulkyQuantifier = new BulkyQuantifier("?", "{0,1}");
+    } else if (Pattern.matches("\\{0,}\\??", raw)) {
+      bulkyQuantifier = new BulkyQuantifier("*", "{0,}");
+    } else if (Pattern.matches("\\{1,}\\??$", raw)) {
```

Comment:
I don't get the optional question mark at the end - even more so because /X+?/ is not equivalent to /X+/ but to /X*/
Is it there to designate a reluctant match?

Reply:
It's a lazy quantifier with the question mark
See https://www.regular-expressions.info/repeat.html#lazy